### PR TITLE
blockchain: Harden and optimize utxo cache.

### PR DESCRIPTION
**Pull request by davecgh**
~**This requires #2994**.~

This reworks several portions of the utxo cache to improve its robustness, optimize it, and correct some hard to hit corner cases that involve a mix of manual block invalidation, conditional flushing, and successive unclean shutdowns along with reworking several of the tests and significantly increasing the associated scenarios covered by the tests.

Note that since **this is consensus critical code**, significant effort has been put into making these changes easier to reason about and review by carefully crafting a series of individual commits such that every commit message thoroughly describes its purpose, maintains consensus, and therefore passes all tests.

The following is a high-level overview of the changes:
- Miscellaneous consistency cleanup
- Pre-allocates the map used to track in-flight transactions in utxo views
- Removes exported utxo cache `SpendEntry` method
- Remove exported utxo cache `AddEntry` method
- Removes unused utxo cache add entry error
- Removes the unnecessary `tip` parameter from the utxo cache `Initialize` method
- Modifies the utxo cache tests to allow overriding cache flushing
- Improves utxo cache initialize tests to ensure initialize code itself flushes the expected data to the backend
- Corrects an extremely hard to hit corner case involving a mix of manual block invalidation, conditional flushing, and successive unclean shutdowns
  - Adds tests for this condition to ensure proper behavior
- Modifies the code that deals with fetching input utxos to avoid needlessly looking up inputs for `treasurybase` and `treasury spend` transactions 
- Updates the code when connecting transactions in a utxo view to avoid adding `treasurybase` utxos since they are never directly spendable
- Simplifies the various utxo cache tests and improves their legibility 
- Reworks utxo cache spend entry tests to simplify them and make them more flexible
- Reworks utxo cache commit tests to simplify them, make them more flexible, and test a lot more conditions
- Improves the robustness of `Commit` in regards to unmodified spent and unspent fresh view entries to help defend against code changes that could easily inadvertently violate the non-obvious, and heretofore undocumented, assumptions
- Reworks the utxo cache add entry tests to simplify them and make them more flexible
- Separates utxo cache and view state by updating the `addEntry` and `fetchEntry` methods to ensure the modified and spent flags are cleared on cloned entries loaded from the cache and and always set correctly when adding entries to the cache
- Modifies the utxo cache commit logic to ensure more robust handling of spent outputs including checking the backend for entries not currently in the cache and conditionally loading them if needed
  - Includes the addition of a double spend assertion
  - Adds tests to ensure the new behavior works as intended
- Splits the logic for handling the connection of transactions from the regular tree and stake tree to a utxo view to provide a clean path for optimizing zero confirmation spends
- Updates the utxo view and cache handling to detect zero confirmation spends in the regular transaction tree and bypass the cache when committing the view to avoid cache misses given they are guaranteed to never have any cache entries since they never exist beyond the scope of the view

### CODE REVIEW

File: internal/blockchain/utxocache.go [hunk 1]
```diff
@@ -314,17 +314,91 @@ func (c *UtxoCache) addEntry(outpoint wire.OutPoint, entry *UtxoEntry) {
...
-	// Attempt to get an existing entry from the cache.
-	cachedEntry := c.entries[outpoint]
+func (c *UtxoCache) spendEntry(outpoint wire.OutPoint) error {
+	// Attempt to get the entry associated with the output from the cache.
+	//
+	// In the case there is a nil entry in the cache for the output, it does not
+	// exist in the backend.  This is because either an attempt to load the utxo
+	// from the backend was made and the nil entry was added to the cache to
+	// avoid further lookups against the backend when the output is known to not
+	// exist or the output was added the the cache and spent in between flushes
```
Comment (chappjc) [inline internal/blockchain/utxocache.go:333]: "the the"

File: internal/blockchain/utxoviewpoint.go [hunk 2]
```diff
@@ -244,41 +259,92 @@ func (view *UtxoViewpoint) connectTransaction(tx *dcrutil.Tx, blockHeight int64,
...
 		// Ignore stakebase since it has no input.
 		if isVote && txInIdx == 0 {
 			continue
 		}
 
-		// Ignore treasury spends since they have no inputs.
-		if isTreasurySpend {
-			continue
+		// Ensure the referenced utxo exists in the view.  This should never
+		// happen unless there is a bug is introduced in the code.
```
Comment (chappjc) [inline internal/blockchain/utxoviewpoint.go:269]: "there is a bug is"

File: internal/blockchain/utxoviewpoint.go [hunk 3]
```diff
@@ -216,26 +216,41 @@ func (view *UtxoViewpoint) PriorityInput(prevOut *wire.OutPoint) (int64, int64,
...
+// is not nil, it will be updated to append an entry for each spent txout.  An
+// error will be returned if the view does not contain the required utxos.
+func (view *UtxoViewpoint) connectStakeTransaction(tx *dcrutil.Tx,
+	blockHeight int64, blockIndex uint32, stxos *[]spentTxOut,
+	isTreasuryEnabled bool) error {
+
+	// Treasurybase transactions don't have any inputs to spend or outputs to
+	// add.
 	msgTx := tx.MsgTx()
 	if isTreasuryEnabled && standalone.IsTreasuryBase(msgTx) {
```
Comment (chappjc) [inline internal/blockchain/utxoviewpoint.go:247]: I'm curious about the use of `IsTreasuryBase` instead of `blockIndex==0`.  Seems like more work than needed.

File: internal/blockchain/utxoviewpoint.go [hunk 4]
```diff
@@ -351,27 +358,42 @@ func (view *UtxoViewpoint) connectRegularTransaction(tx *dcrutil.Tx,
...
 	return nil
 }
 
 // connectRegularTransactions updates the view by adding all new utxos created
 // by the transactions in the regular transaction tree of the block and marking
-// all utxos those same transactions spend as spent.  In addition, when the
-// 'stxos' argument is not nil, it will be updated to append an entry for each
-// spent txout.  An error will be returned if the view does not contain the
-// required utxos.
+// all utxos those same transactions spend as spent.  It also marks all earlier
```
Comment (dnldd) [inline internal/blockchain/utxoviewpoint.go:382]: s / `...and marking all utxos those same transactions spend as spent.` /  `...and marking all utxos with the same transactions spend as spent.`

File: internal/blockchain/utxocache.go [hunk 5]
```diff
@@ -451,15 +451,15 @@ func (c *UtxoCache) Commit(view *UtxoViewpoint) error {
 			continue
 		}
```
Comment (dnldd) [inline internal/blockchain/utxocache.go:559]: there are a couple of typos in the commit message here: 
- s / `... It accomplishes this by introdcuing an ...`/ `It accomplishes this by introducing an ...`
- s / `... the non-obvious, and heretofore undocumented, ...`, / `the non-obvious, and therefore undocumented, ...`

**Comment by dnldd** [comment]
Looks good overall, will be running a mainnet sync with this soon.

File: internal/blockchain/utxoviewpoint.go [hunk 6]
```diff
@@ -755,8 +877,16 @@ func (view *UtxoViewpoint) fetchInputUtxos(block *dcrutil.Block,
 	// the block.  This applies to both transactions earlier in the stake tree
 	// as well as those in the regular tree.
 	for _, stx := range block.STransactions() {
-		isVote := stake.IsSSGen(stx.MsgTx())
-		for txInIdx, txIn := range stx.MsgTx().TxIn {
+		// Ignore treasurybases and treasury spends since they have no inputs.
+		msgTx := stx.MsgTx()
+		isTreasuryBase := isTreasuryEnabled && stake.IsTreasuryBase(msgTx)
+		isTreasurySpend := isTreasuryEnabled && stake.IsTSpend(msgTx)
+		if isTreasuryBase || isTreasurySpend {
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go:884]: I'm a bit torn on the first one to be honest.  I did consider it for the first case, because the stake identification functions are indeed not very efficient currently, but I was a bit hesitant because solely relying on the index would introduce an implicit ordering assumption that code prior to this has already validated that the first transaction is actually the required `treasurybase` and the rest are not.  That is a true assumption in this particular case as well as the other one mentioned in on of your other comments.

In general, I prefer to avoid as many hidden implicit ordering assumptions like that as possible in consensus except in cases where it really has a huge impact on performance, but I do believe this is a good candidate given it applies to every single stake transaction and the stake identification functions currently do quite a bit of work.

That said, I would say it should be in addition to the check as opposed to in lieu of it (so that it only checks for the first transaction) to act as an additional bulwark.  In order words, this seems reasonable to cover the bases:
```Go
isTreasuryBase := isTreasuryEnabled && txIdx == 0 && stake.IsTreasuryBase(msgTx)
```

For the second case, I did consider that, as it's done in some other parts of the code (e.g. `mempool`) but I decided against it here since unlike the other cases, it necessarily could only realistically possibly avoid the check on a single transaction here versus adding another conditional branch for every other transaction.  We'd need to benchmark it to really see if it's worth it, but my gut tells me it would probably be a wash and ultimately with more efficient stake identification would likely be worse.

Sound reasonable?

File: internal/blockchain/utxoviewpoint.go (hunk 3 above)
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go:247]: Discussed in the other related comment.

File: internal/blockchain/utxoviewpoint.go (hunk 4 above)
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go:382]: I don't think that would be correct.  Spend is the verb while spent is the noun.  So, the existing "marking all utxos those same transaction spend (the verb) as spent (the noun)" seems accurate to me.

That said, I'm not opposed to coming up from different wording if you find the way it is currently written to be hard to parse.

File: internal/blockchain/utxocache.go (hunk 5 above)
Comment (davecgh) [reply internal/blockchain/utxocache.go:559]: Fixed the first case.

The second one is correct.  [heretofore](https://www.merriam-webster.com/dictionary/heretofore) means "up to this time".

File: internal/blockchain/utxocache.go [hunk 7]
```diff
@@ -314,17 +314,91 @@ func (c *UtxoCache) addEntry(outpoint wire.OutPoint, entry *UtxoEntry) {
...
+	// cache flush in the case it exists, the associated cache entry from the
+	// backend is loaded and marked spent.
+	if !found {
+		// Account for the cache miss and attempt to load the entry for the utxo
+		// from the backend.
+		//
+		// Notice that this intentionally attempts to load the entry from the
+		// backend directly as opposed to using the normal cache entry fetching
+		// method since that involves querying the cache again which would be
+		// wasteful since it is already known to not exist as well as slightly
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxocache.go]: Updated.

File: internal/blockchain/utxocache.go (hunk 1 above)
Comment (davecgh) [reply internal/blockchain/utxocache.go]: Updated.

File: internal/blockchain/utxoviewpoint.go (hunk 2 above)
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go]: Modified.

File: internal/blockchain/utxoviewpoint.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (chappjc) [reply internal/blockchain/utxoviewpoint.go:884]: Yeah.  Although for the second case, I think it would be something minor like the following short-circuit:

```go
isTreasurySpend := isTreasuryEnabled && !isTreasuryBase && stake.IsTSpend(msgTx)
```

But it only saves a few microseconds for the case where it is the treasurybase.

[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go:884]: I went ahead and modified it to the following which covers both cases, I believe makes the assumption more clear, and avoid the additional conditional branch that would be introduced by checking it twice:

```Go
		shouldBeTreasuryBase := isTreasuryEnabled && txIdx == 0
		if shouldBeTreasuryBase && stake.IsTreasuryBase(msgTx) {
			continue
		}
		isTreasurySpend := isTreasuryEnabled && stake.IsTSpend(msgTx)
		if isTreasurySpend {
			continue
		}
```

File: internal/blockchain/utxoviewpoint.go (hunk 3 above)
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go:247]: I updated this as follows:

```Go
	//
	// NOTE: This check relies the code already having validated that the first
	// transaction, and only the first transaction, in the stake tree is the
	// required treasurybase when the treasury agenda is active.
	msgTx := tx.MsgTx()
	if isTreasuryEnabled && blockIndex == 0 {
```

It's a lot safer to rely on this assumption here than the other case, because this case is required to be much deeper in the consensus checks which means it is exceedingly unlikely the code could be changed in such a way as to violate it.

File: internal/blockchain/utxoentry.go [hunk 8]
```diff
@@ -149,6 +154,13 @@ func (entry *UtxoEntry) isFresh() bool {
 	return entry.state&utxoStateFresh == utxoStateFresh
 }
 
+// isSpentByZeroConf returns whether or not the output is marked as spent by a
+// zero confirmation transaction.  In other words, it is an output that was
+// create in a block and later spent by another transaction in the same block.
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxoentry.go]: Updated.

File: internal/blockchain/utxoviewpoint.go (hunk 4 above)
Comment (dnldd) [reply internal/blockchain/utxoviewpoint.go:386]: You're right, it's clearer after a second read. Disregard.

**Comment by dnldd** [comment]
Synced mainnet with the changes made here without issue.

**Review by dnldd — APPROVED** [review]

File: internal/blockchain/validate.go [hunk 9]
```diff
@@ -3515,16 +3519,24 @@ func (b *BlockChain) checkTransactionsAndConnect(inputFees dcrutil.Amount, node
...
-		if err != nil {
-			return err
+		if !stakeTree {
+			err := view.connectRegularTransaction(tx, node.height, uint32(idx),
+				inFlightRegularTx, stxos, isTreasuryEnabled)
+			if err != nil {
+				return err
+			}
+		} else {
+			err = view.connectStakeTransaction(tx, node.height, uint32(idx),
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/validate.go]: Done.

**Review by JoeGruffins — APPROVED** [review]

File: internal/blockchain/utxoviewpoint.go [hunk 10]
```diff
@@ -216,26 +216,45 @@ func (view *UtxoViewpoint) PriorityInput(prevOut *wire.OutPoint) (int64, int64,
...
+	blockHeight int64, blockIndex uint32, stxos *[]spentTxOut,
+	isTreasuryEnabled bool) error {
+
+	// Treasurybase transactions don't have any inputs to spend or outputs to
+	// add.
 	//
-	// NOTE: This check MUST come before the coinbase check because a
-	// treasurybase is identified as a coinbase as of the time this comment is
-	// being written.
+	// NOTE: This check relies the code already having validated that the first
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxoviewpoint.go]: Updated.

File: internal/blockchain/utxocache.go [hunk 11]
```diff
@@ -314,17 +314,91 @@ func (c *UtxoCache) addEntry(outpoint wire.OutPoint, entry *UtxoEntry) {
...
-	// Attempt to get an existing entry from the cache.
-	cachedEntry := c.entries[outpoint]
+func (c *UtxoCache) spendEntry(outpoint wire.OutPoint) error {
+	// Attempt to get the entry associated with the output from the cache.
+	//
+	// In the case there is a nil entry in the cache for the output, it does not
+	// exist in the backend.  This is because either an attempt to load the utxo
+	// from the backend was made and the nil entry was added to the cache to
+	// avoid further lookups against the backend when the output is known to not
+	// exist or the output was added to the cache and spent in between flushes
```
[the remark this answers is not in the document]
Comment (davecgh) [reply internal/blockchain/utxocache.go:333]: Correct.  That is done elsewhere.  Notice the comment below calls that out:

```Go
		//
		// Note that a nil entry is not added to cache for the output here since
		// there should realistically never be any future lookups for the spent
		// output given that would be a double spend.
```
